File: functional-error-handling/code/Downloader.Api/Controllers/JohnAmountController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Downloader.Api.Other;
using Downloader.Api.Shared;
using Microsoft.AspNetCore.Mvc;

namespace Downloader.Api.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    public class JohnAmountController : ControllerBase
    {
        [HttpGet("{fileName}")]
        public ActionResult Get(string fileName)
        {
            var downloader = new Downloader();
            var parser = new Parser();

            try
            {
                var content = downloader.GetFile(fileName);

                try
                {
                    var csvLines = parser.Parse(content);

                    var amount = csvLines
                        .Where(x => x.Name == "John")
                        .Sum(x => x.Amount);

                    return Ok(amount);
                }
                catch (FileDoesntParseException)
                {
                    return UnprocessableEntity();
                }
            } catch (SftpUnauthorisedException)
            {
                return Unauthorized();
            }
            catch (FileDoesntExistException)
            {
                return NotFound();
            }
        }

        public class SftpUnauthorisedException : Exception { }
        public class FileDoesntExistException : Exception { }
        public class FileDoesntParseException : Exception { }

        public class Downloader
        {
            public string GetFile(string fileName)
            {
                if (fileName == FileNames.UnauthorisedSftp)
                {
                    throw new SftpUnauthorisedException();
                }

                if (fileName == FileNames.FileMissingOnSftp)
                {
                    throw new FileDoesntExistException();
                }

                if (fileName == FileNames.FileDoesntParse)
                {
                    return FileContent.InvalidFile;
                }

                return FileContent.ValidFile;
            }
        }

        public class Parser
        {
            public IEnumerable<PersonAmount> Parse(string content)
            {
                var fileParser = new FileParser();

                try
                {
                    return fileParser.Parse(content);
                }
                catch (Exception)
                {
                    throw new FileDoesntParseException();
                }
            }
        }
    }
}

File: functional-error-handling/code/Downloader.Api.Fs/Controllers/JohnAmountController.fs
namespace Downloader.Api.Fs.Controllers

open System
open Microsoft.AspNetCore.Mvc

module JohnAmount =
    open Downloader.Api.Shared

    type DownloadingError = SftpUnauthorized | FileNotPresent
    
    type JohnAmountError = 
        | DownloadingError of DownloadingError
        | FileDoesNotParse
    
    type PersonAmount = { Name: string; Amount: decimal }

    let download fileName =
        match fileName with
        | FileNames.UnauthorisedSftp -> Error SftpUnauthorized
        | FileNames.FileMissingOnSftp -> Error FileNotPresent
        | FileNames.FileDoesntParse -> Ok FileContent.InvalidFile
        | _ -> Ok FileContent.ValidFile
    
    let parse (content: string) =
        let parseLine (line: string) = 
            match line.Split(',') with
            | [| name; amount |] -> 
                match Decimal.TryParse amount with
                | (true, result) -> Some { Name = name ; Amount = result }
                | (false, _) -> None
            | _ -> None
        
        content.Split('\n') 
        |> Array.fold (fun agg el ->
            let parsed = parseLine el
            match parsed with
            | Some personAmount -> agg |> Result.map (fun lines -> lines @ [personAmount])
            | None -> Error FileDoesNotParse
        ) (Ok [])


open JohnAmount

[<Route("api/[controller]")>]
[<ApiController>]
type JohnAmountController () =
    inherit ControllerBase()

    [<HttpGet("{fileName}")>]
    member this.Get(fileName:string) =
        
        download fileName |> Result.mapError DownloadingError
        |> Result.bind parse
        |> Result.map (fun personAmounts ->
            personAmounts 
            |> List.filter (fun x -> x.Name = "John")
            |> List.sumBy (fun x -> x.Amount)
        )
        |> (fun res ->
                match res with 
                | Ok amount -> JsonResult(amount) :> ActionResult
                | Error error ->
                    match error with
                    | DownloadingError downloadingError -> 
                        match downloadingError with
                        | SftpUnauthorized -> this.Unauthorized() :> ActionResult
                        | FileNotPresent -> this.NotFound() :> ActionResult
                    | FileDoesNotParse -> this.UnprocessableEntity() :> ActionResult
        )
